### PR TITLE
Ensure pointer invariants are maintained

### PR DESCRIPTION
**Pull request by reuk**
It used to be fine to create a `pointer_typet` from a `typet`, but it's not any more. Unfortunately, it's not possible to use the type system to enforce this, which means that we'll have to wait until the `asserts` I've added get violated to (maybe) find all the bad conversions.

@allredj this should fix the models-library bug.
@LAJW, could you review too?

### CODE REVIEW

**Comment by LAJW** [comment]
Keep in mind that AppVeyor compiles Release build, which disables asserts, so in case CI fails, that's where you might want to look.

**Review by LAJW — APPROVED** [review]

**Comment by reuk** [comment]
There's some issue in some of the string solver regression tests on Travis. Maybe @allredj or @romainbrenguier could take a look at the log to see if it's a known error?

**Comment by allredj** [comment]
@reuk I'm yet to find a single test that passes on that build...

**Comment by reuk** [comment]
Wonderful

**Comment by reuk** [comment]
OK, I'll take a look at this again. Could someone else update the cbmc submodule? I tried, but it took me an hour to get a compiling version of test-gen, and *that* fails most regression tests too.

**Comment by reuk** [comment]
I ran *all* the regression tests locally this time, and they seem to pass. The unit tests still do not pass, though.

**Comment by reuk** [comment]
Looks like this will pass CI now. Please re-review

**Comment by allredj** [comment]
So what's with the failing AppVeyor build?

**Comment by reuk** [comment]
I'm not sure... I don't understand why it would work on Travis but not Appveyor. I'll look into it

**Comment by reuk** [comment]
The head of CBMC develop is broken, because failing tests were enabled. Once #1377 is merged (or the failing tests are fixed) I can rebase onto a working version of master, and then we can commit.

**Review by allredj — APPROVED** [review]
Seems alright to me, although I'm not familiar with that code.
The AppVeyor problem should be fixed by #1377.

**Comment by reuk** [comment]
Rebased. Hopefully this will pass CI now.

**Comment by reuk** [comment]
@peterschrammel please review and merge (if acceptable)

File: src/ansi-c/c_typecheck_type.cpp [hunk 1]
```diff
@@ -1492,9 +1492,7 @@ void c_typecheck_baset::adjust_function_parameter(typet &type) const
 {
   if(type.id()==ID_array)
   {
-    type.id(ID_pointer);
-    type.remove(ID_size);
-    type.remove(ID_C_constant);
+    type=pointer_type(type.subtype());
```
Comment (peterschrammel) [inline src/ansi-c/c_typecheck_type.cpp]: I assume you have checked that this is indeed equivalent.

File: src/util/std_types.h [hunk 2]
```diff
@@ -1391,6 +1391,7 @@ class pointer_typet:public bitvector_typet
 inline const pointer_typet &to_pointer_type(const typet &type)
 {
   assert(type.id()==ID_pointer);
+  assert(!type.get(ID_width).empty());
```
Comment (peterschrammel) [inline src/util/std_types.h]: It would be good to move this over to invariant.h

**Comment by reuk** [comment]
@peterschrammel requested changes are in place

**Pull request merged**